src/util/parallel/data.ml

```diff
@@ -0,0 +1,242 @@
+        let new_sll = {
+          key = key;
+          value = value;
+          next = Atomic.make None;
+        } in
+        let success = Atomic.compare_and_set sll.next None (Some new_sll) in
+        if not success then insert_value sll key value
+      | Some next -> insert_value next key value
+    )
+
+  let to_list sll =
+    let rec aux acc sll =
+      match sll with
+      | None -> acc
+      | Some sll ->
+        aux (acc @ [(sll.key, sll.value)]) (Atomic.get sll.next)
+    in
+    aux [] (Some sll)
+
+  let to_seq sll =
+    let rec aux sll () =
+      match sll with
+      | None -> Seq.Nil
+      | Some sll -> Seq.Cons ((sll.key, sll.value), aux (Atomic.get sll.next))
+    in aux (Some sll)
+
+  let to_string sll =
+    let rec aux acc sll =
+      match sll with
+      | None -> acc
+      | Some sll ->
+        aux (acc ^ Val.to_string (Atomic.get sll.value) ^ " ") (Atomic.get sll.next)
+    in
+    aux "" (Some sll)
+end
+
+
+(* This is a custom implementation, because we leave out operations
+    that we do not need to enable a more efficient implementation. *)
+module ConcurrentHashmap =
+  functor (H: Hashtbl.HashedType) ->
+  functor (D: DefaultType) ->
+  functor (HM:Hashtbl.S with type key = H.t) ->
+  struct
+    module Bucket = ConcurrentBucket(H)(D)
+
+    type key = H.t
+    type value = D.t Atomic.t
+
+    type t = {
+      size: int Atomic.t;
+      nr_elements: int Atomic.t;
+      resize_generation: int Atomic.t;
+      buckets: Bucket.t option Atomic.t array Atomic.t;
+    }
+
+    let create () = 
+      let size = 100 in
+      {
+        size = Atomic.make size;
+        nr_elements = Atomic.make 0;
+        resize_generation = Atomic.make 0;
+        buckets = Atomic.make @@ Array.init size (fun _ -> Atomic.make None);
+      }
+
+    let to_list hm =
+      Array.fold_left (fun acc bucket ->
+          match Atomic.get bucket with
+          | None -> acc
+          | Some bucket -> acc @ Bucket.to_list bucket
+        ) [] (Atomic.get hm.buckets)
+
+    let to_seq hm =
+      let bucket_seq = Array.to_seq (Atomic.get hm.buckets) in
+      let non_atomic_bucket_seq = Seq.map Atomic.get bucket_seq in
+      let non_option_bucket_seq = Seq.filter (fun x -> x <> None) non_atomic_bucket_seq in
+      let non_atomic_bucket_seq = Seq.map (fun x -> match x with | Some x -> x | None -> failwith "This should not happen") non_option_bucket_seq in
+      Seq.flat_map Bucket.to_seq non_atomic_bucket_seq
+
+    let find_option hm key =
+      let hash = abs @@ H.hash key in
+      let bucket = Array.get (Atomic.get hm.buckets) (hash mod (Atomic.get hm.size)) in
+      match Atomic.get bucket with
+      | None -> None
+      | Some bucket -> Bucket.find_option bucket key
+
+    let find hm key =
+      match find_option hm key with
+      | None -> failwith "Key not found"
+      | Some value -> value
+
+    let mem hm key =
+      match find_option hm key with
+      | None -> false
+      | Some _ -> true
```

Comment:
```suggestion
     Option.is_some @@ find_option hm key
```

maybe even 

~~~Ocaml
let mem = Option.is_some @@ find_option
~~~

Never know when eta expansion is needed...